### PR TITLE
Refactor structure

### PR DESCRIPTION
**Pull request by Borda**
creating a single package and moving all training/predict script to a single folder

### CODE REVIEW

**Comment by jamt9000** [comment]
Hello, thanks for the PR!

Since this is a relatively large set of changes could you give some more details and motivation about your proposal?

I think it generally looks good and a package vs training split is cleaner. But for example this would move the data_loaders to `detoxify/detoxify/utils` when currently `detoxify/detoxify` is just for the code that ends up in the installed package.

File: scripts/compute_bias_metric.py [hunk 1]
```diff
@@ -3,7 +3,18 @@
 import numpy as np
 from sklearn.metrics import roc_auc_score
 import argparse
-from utils import compute_auc, compute_subgroup_auc
+
+
+def compute_auc(y_true, y_pred):
```
Comment (jamt9000) [inline scripts/compute_bias_metric.py:8]: This is now defined twice (also in compute_language_breakdown.py) so should probably keep it in one place

Comment (Borda) [reply scripts/compute_bias_metric.py:8]: well yes, but then I would consider moving it to the detoxify package, as having it here in script won't work as it is not a package available for import (missing `__init__.py` file and I would keep it such)

**Comment by Borda** [comment]
> Since this is a relatively large set of changes could you give some more details and motivation about your proposal?

In the beginning, I wanted just to expose some arguments for offline prediction, but then I found myself a bit hard to navigate what is what... so I started some cleaning in the sense make it a single package (now you mix `detoxify` package and some other is in src folder) scripts are also on multiple places...

> But for example this would move the data_loaders to `detoxify/detoxify/utils` when currently `detoxify/detoxify` is just for the code that ends up in the installed package.

I think it is completely fine to include dataloaders in the package so a user can build on top of it or extend it

File: .github/workflows/ci-testing.yml [hunk 2]
```diff
@@ -3,12 +3,9 @@ name: CI testing
 # see: https://help.github.com/en/actions/reference/events-that-trigger-workflows
 on:
   # Trigger the workflow on push or pull request, but only for the master branch
-  push:
-    branches:
-      - master
+  push: {}
```
Comment (Borda) [inline .github/workflows/ci-testing.yml:6]: this help run CI even for branches before you create a PR

File: .github/workflows/ci-testing.yml [hunk 3]
```diff
@@ -53,16 +50,16 @@ jobs:
 
     - name: Install dependencies
       run: |
-        pip install --requirement requirements.txt --upgrade --quiet --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --use-feature=2020-resolver
-        pip install --requirement tests/requirements.txt --quiet --use-feature=2020-resolver
+        pip install --requirement requirements.txt --upgrade --quiet --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
+        pip install --requirement tests/requirements.txt --quiet
```
Comment (Borda) [inline .github/workflows/ci-testing.yml:54]: resolver is not needed anymore

File: detoxify/core.py [hunk 4]
```diff
@@ -1,12 +1,13 @@
 import torch
 import transformers
 
+DOWNLOAD_URL = "https://github.com/unitaryai/detoxify/releases/download/"
```
Comment (Borda) [inline detoxify/core.py:4]: this is mainly to make the lines shorted and so please flake8 and 120 line length

**Comment by jamt9000** [comment]
> I think it is completely fine to include dataloaders in the package so a user can build on top of it or extend it

The issue here is that those files have a lot of dependencies (sklearn, pandas, tqdm, datasets) which are not needed for making predictions and are not included in setup.py or the pypi package at the moment (and I think most users of the pypi package users just want to do inference so we shouldn't require these large train-time dependencies).

You can test how the installed package would work with:

`pip install 'git+https://github.com/Borda/detoxify@refactor/struct'`

Currently for training we expect people to clone the repo and install `requirements.txt` so that doesn't really involve the `detoxify` package.

**Comment by Borda** [comment]
> Currently for training we expect people to clone the repo and install `requirements.txt` so that doesn't really involve the `detoxify` package.

interesting, and for this case how about having the train dependencies in extras?
so a user for inference would call `pip install detoxify` and if with training use `pip install detoxify[train]`

**Comment by Borda** [comment]
@laurahanu may I get your review also on this one? :rabbit:

**Comment by laurahanu** [comment]
Thanks for this PR, I agree we could benefit from some restructuring. Since the data loaders are meant to be used with the training scripts provided and would need to be changed for other datasets, I think it makes sense to have them outside the package so that they can be easily modified.
So maybe could have something like:

 ┣ configs
 ┣ detoxify
 ┃ ┣ `__init__.py`
 ┃ ┗ `detoxify.py`
 ┃ ┗ `utils.py`
 ┣ training
 ┃ ┣ `data.py`
 ┃ ┣ `preprocessing_utils.py`
 ┃ ┗ `train.py`
 ┣ evaluation
 ┃ ┣ `bias_metrics.py`
 ┃ ┣ `compute_bias_metric.py`
 ┃ ┣ `compute_language_breakdown.py`
 ┃ ┗ `evaluate.py`
 ┣ scripts
 ┃ ┣ `convert_weights.py`
 ┃ ┗ `run_prediction.py`
 ┣ tests

**Comment by Borda** [comment]
> I think it makes sense to have them outside the package so that they can be easily modified.

what do you mean? you can edit anything also inside the package even easier as you have it accessible under a version

**Comment by jamt9000** [comment]
> what do you mean? you can edit anything also inside the package even easier as you have it accessible under a version

This tends to cause confusion and conflicts between the `site-packages` installed version of the library and the one in the source tree (Such as https://github.com/pytorch/vision/issues/2239).

**Comment by Borda** [comment]
@jamt9000 so you suggest dropping this PR? I am fine with it, so will continue in clean code in the my fork... :rabbit:

**Comment by Borda** [comment]
closing this PR as there is no will to improve project structure... :rabbit:

**Pull request closed without merging**